src/Transports.AspNetCore/WebSockets/IWebSocketConnection.cs

```diff
@@ -0,0 +1,50 @@
+using GraphQL.Transport;
+using Microsoft.AspNetCore.Http;
+
+namespace GraphQL.Server.Transports.AspNetCore.WebSockets;
+
+/// <summary>
+/// Represents a WebSocket connection, dispatching received messages over
+/// the connection to the specified <see cref="IOperationMessageProcessor"/>,
+/// and sending requested messages out the connection when requested.
+/// <br/><br/>
+/// <see cref="ExecuteAsync(IOperationMessageProcessor)"/> may be only called once.
+/// All members must be thread-safe.
+/// </summary>
+public interface IWebSocketConnection
+{
+    /// <summary>
+    /// Listens to incoming messages over the WebSocket connection,
+    /// dispatching the messages to the specified <paramref name="operationMessageProcessor"/>.
+    /// Returns or throws <see cref="OperationCanceledException"/> when the WebSocket connection is closed.
+    /// </summary>
+    Task ExecuteAsync(IOperationMessageProcessor operationMessageProcessor);
+
+    /// <summary>
+    /// Sends a message.
+    /// </summary>
+    Task SendMessageAsync(OperationMessage message);
+
+    /// <summary>
+    /// Closes the WebSocket connection.
+    /// </summary>
+    Task CloseConnectionAsync();
```

Comment:
`IOperationMessageProcessor` implements `IDisposable` but here you use dedicated method to close connection. What about `Dispose` here?

Reply:
I guess this method should send special message for client. So what about lifetime of connection after this method being called?

Reply:
`CloseConnectionAsync` posts a WebSocket message through the message pump to the client that informs it that the WebSocket connection is shutting down.  Per HTTP WebSocket protocol, a graceful shutdown occurs only after both endpoints have sent/acknowledged a close message.  Per GraphQL over WebSocket protocols, the close event id needs to be certain values for certain events.

`CloseConnectionAsync` does not dispose of the `IWebSocketConnection` implementation, as it must stay alive until any remaining messages in its queue have sent and the counterparty has responded with a close message.  In fact, there is no `Dispose` method at all because all of the held objects (e.g. `WebSocket`) originate from `HttpContext` and are disposed by ASP.NET Core.

I did just now add `IDisposable` inheritance to `IWebSocketConnection` anyway, in case of code changes in the future or other implementations, although it does not contain any code at this time.

Within the implementation (#809 ), `ExecuteAsync` will return once the graceful shutdown has completed.  Once the middleware returns, ASP.NET Core will close the HTTP connection.

Reply:
Thanks. And again I note that I would happy to see these explanations in the xml comments.